File: frontend/lib/report-values.ts
// ABOUTME: 报告级结构化事实的前端派生与引用解析，镜像后端 contract/report_values.py。
// ABOUTME: 仅从 Report 读取展示值，不维护副本，不参与 prompt 拼装。
import { selectedStandardNames } from "./disclosure-basis";
import { listSeparatorFor, reportLanguageFor } from "./package-language";
import type { Report } from "./schema";

const REPORT_REF_LABELS: Record<string, string> = {
  "disclosureProfile.basisStatement": "报告编制依据",
  "disclosureProfile.selectedStandardNames": "适用披露准则",
  "appendixPackage.externalAssuranceReport.fileLabel": "外部鉴证报告文件",
  "appendixPackage.readerFeedbackContactInformation.address": "读者反馈通信地址",
  "appendixPackage.readerFeedbackContactInformation.email": "读者反馈邮箱",
  "appendixPackage.readerFeedbackContactInformation.phone": "读者反馈电话",
  "assessment.counts.dual": "双重重要性议题数量",
  "assessment.counts.impact_only": "影响重要性议题数量",
  "assessment.counts.financial_only": "财务重要性议题数量",
  "assessment.counts.non_material": "非重要性议题数量",
  "assessment.applicableTopicCount": "本报告披露议题数量",
};

export function selectedStandardNamesText(report: Report): string {
  return selectedStandardNames(report).join(listSeparatorFor(report.knowledgePackageId));
}

export function disclosureBasisStatement(report: Report): string {
  return selectedStandardNamesText(report);
}

function appendixValue(path: string, report: Report): unknown {
  const pkg = report.appendixPackage;
  if (path === "appendixPackage.externalAssuranceReport.isIncluded") return pkg?.externalAssuranceReport?.isIncluded ?? false;
  if (path === "appendixPackage.externalAssuranceReport.fileLabel") {
    return pkg?.externalAssuranceReport?.fileLabel ?? null;
  }
  if (path === "appendixPackage.readerFeedbackContactInformation.address") return pkg?.readerFeedbackContactInformation?.address ?? null;
  if (path === "appendixPackage.readerFeedbackContactInformation.email") return pkg?.readerFeedbackContactInformation?.email ?? null;
  if (path === "appendixPackage.readerFeedbackContactInformation.phone") return pkg?.readerFeedbackContactInformation?.phone ?? null;
  return null;
}

function disclosureValue(path: string, report: Report): unknown {
  const profile = report.disclosureProfile;
  if (path === "disclosureProfile.mainlandStandard") return profile?.mainlandStandard ?? "sse";
  if (path === "disclosureProfile.includesHongKongExchangeGuide") return profile?.includesHongKongExchangeGuide ?? false;
  if (path === "disclosureProfile.additionalDisclosureReferences") return profile?.additionalDisclosureReferences ?? [];
  if (path === "disclosureProfile.selectedStandardNames") return selectedStandardNamesText(report);
  if (path === "disclosureProfile.basisStatement") return disclosureBasisStatement(report);
  return null;
}

export function resolveReportRef(ref: string | null | undefined, report: Report): unknown {
  if (!ref) return null;
  if (ref.startsWith("disclosureProfile.")) return disclosureValue(ref, report);
  if (ref.startsWith("appendixPackage.")) return appendixValue(ref, report);
  // assessment.* 由 report-document/document-nodes 的 assessmentValue 现算：议题范围是
  // 知识包属性（上交所 24 个、港交所 16 个），任何前端基数都必然对某个包写错。
  return report.fields[ref]?.value ?? null;
}

/** 报告正文中的空引用使用业务名称占位，绝不把内部路径暴露给用户。 */
export function reportRefLabel(ref: string | null | undefined, report: Report): string {
  if (!ref) return "待填写信息";
  return report.fields[ref]?.label ?? REPORT_REF_LABELS[ref] ?? "待填写信息";
}

/**
 * 报告正文里的引用显示值。
 *
 * 分隔符与布尔字面量都随报告所属知识包的语言：英文正文里的「A、B」与「是」是中文标点和
 * 中文词，不是英文报告该有的形态。
 */
export function reportValueToText(value: unknown, report: Report): string {
  if (Array.isArray(value)) {
    return value
      .map((item) => String(item).trim())
      .filter(Boolean)
      .join(listSeparatorFor(report.knowledgePackageId));
  }
  if (value === null || value === undefined || value === "") return "";
  if (typeof value === "boolean") return booleanText(value, report);
  return String(value);
}

/** 布尔字段在正文中的显示词，按报告所属包的语言。 */
function booleanText(value: boolean, report: Report): string {
  const language = reportLanguageFor(report.knowledgePackageId);
  if (language === "en") return value ? "Yes" : "No";
  if (language === "zh-Hant") return value ? "是" : "否";
  return value ? "是" : "否";
}

File: backend/src/sustainability_desk/contract/report_values.py
# ABOUTME: 报告级结构化事实的确定性派生与引用解析，供诊断、导出与前端镜像实现对齐。
# ABOUTME: 本模块只读取 Report，不接触 LLM prompt，不暴露内部原始字段结构。
# ABOUTME(en): Deterministic derivation of report-level structured facts and reference resolution.
# ABOUTME(en): Reads a Report only; touches no LLM prompt and exposes no internal raw field structure.
from __future__ import annotations

import re

from sustainability_desk.contract.knowledge_packages import knowledge_package_of
from sustainability_desk.contract.language import format_month
from sustainability_desk.contract.models import Report

# 联系邮箱采用 ASCII 商务邮箱语法；与前端 isFieldValueAllowed("email") 经
# tests/fixtures/email_validation_golden.json 对齐，两端不得单独放宽或收紧。
CONTACT_EMAIL_PATTERN = re.compile(r"^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$")


def is_valid_contact_email(value: str) -> bool:
    """判定非空联系邮箱格式；空串代表未填写，由调用方在此之前处理。"""
    return bool(CONTACT_EMAIL_PATTERN.fullmatch(value))


def selected_standard_names(report: Report) -> list[str]:
    """根据报告配置输出对外展示的准则名称列表。

    The package's disclosure basis decides the shape: a selectable mainland basis reads the user's
    choice from ``disclosureProfile``; a fixed basis names its standards outright.
    """
    basis = knowledge_package_of(report).manifest.disclosure_basis
    profile = report.disclosureProfile
    if basis.mainland_standard_selectable:
        mainland = profile.mainlandStandard if profile else "sse"
        names = [basis.mainland_standard_names.name(mainland)]
        if profile and profile.includesHongKongExchangeGuide and basis.hong_kong_guide_name:
            names.append(basis.hong_kong_guide_name)
    else:
        names = list(basis.primary_standard_names)
    for name in profile.additionalDisclosureReferences if profile else []:
        clean = str(name).strip()
        if clean:
            names.append(clean)
    return names


def selected_standard_names_text(report: Report) -> str:
    return "、".join(selected_standard_names(report))


def disclosure_basis_statement(report: Report) -> str:
    return selected_standard_names_text(report)


def _appendix_value(path: str, report: Report):
    pkg = report.appendixPackage
    if pkg is None:
        return None
    if path == "appendixPackage.externalAssuranceReport.isIncluded":
        return pkg.externalAssuranceReport.isIncluded
    if path == "appendixPackage.externalAssuranceReport.fileLabel":
        return pkg.externalAssuranceReport.fileLabel
    if path == "appendixPackage.readerFeedbackContactInformation.address":
        return pkg.readerFeedbackContactInformation.address
    if path == "appendixPackage.readerFeedbackContactInformation.email":
        return pkg.readerFeedbackContactInformation.email
    if path == "appendixPackage.readerFeedbackContactInformation.phone":
        return pkg.readerFeedbackContactInformation.phone
    return None


def _disclosure_value(path: str, report: Report):
    profile = report.disclosureProfile
    if path == "disclosureProfile.mainlandStandard":
        return profile.mainlandStandard if profile else "sse"
    if path == "disclosureProfile.includesHongKongExchangeGuide":
        return profile.includesHongKongExchangeGuide if profile else False
    if path == "disclosureProfile.additionalDisclosureReferences":
        return profile.additionalDisclosureReferences if profile else []
    if path == "disclosureProfile.selectedStandardNames":
        return selected_standard_names_text(report)
    if path == "disclosureProfile.basisStatement":
        return disclosure_basis_statement(report)
    return None


def resolve_report_ref(ref: str | None, report: Report):
    """解析行内引用和值路径；未知引用返回 None，由调用方决定是否占位或诊断。"""
    if not ref:
        return None
    if ref.startswith("disclosureProfile."):
        return _disclosure_value(ref, report)
    if ref.startswith("appendixPackage."):
        return _appendix_value(ref, report)
    if ref == "assessment.applicableTopicCount":
        # 本报告实际披露的议题数量：取「必须由用户评分」的范围，不取含 fixed 的
        # 重要性范围（后者多一个 stakeholder_communication，不进议题章节）。
        # 现算而非读快照——该数随科技伦理适用性与披露档位变化，写死必然写错。
        # 延迟 import：visibility/report_values 刻意保持轻量，不在模块级依赖议题注册表。
        from sustainability_desk.contract.topic_registry import applicable_scoring_topics

        return len(applicable_scoring_topics(report))
    field = report.fields.get(ref)
    if field is not None:
        return field.value
    return None


def display_report_ref(ref: str | None, report: Report):
    """将已解析字段投影为正式报告显示值，不改变 Report 中的原始 typed value。"""

    value = resolve_report_ref(ref, report)
    field = report.fields.get(ref or "")
    if field is not None and field.type == "month" and value not in (None, ""):
        return format_month(str(value), knowledge_package_of(report).language)
    return value
